sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClient.java

```diff
@@ -330,6 +334,10 @@ public final class ServiceBusReceiverAsyncClient implements AutoCloseable {
     private static final Duration EXPIRED_RENEWAL_CLEANUP_INTERVAL = Duration.ofMinutes(2);
     private static final DeadLetterOptions DEFAULT_DEAD_LETTER_OPTIONS = new DeadLetterOptions();
     private static final String TRANSACTION_LINK_NAME = "coordinator";
+    /**
+     * The maximum number of messages to delete in a single batch.  This cap is established and enforced by the service.
+     */
+    static final int MAX_DELETE_MESSAGES_COUNT = 4000;
```

Comment:
We try not to set these constants in case the service changes the cap. in real time... folks would need to update the SDK to get around this.  It would be preferred to send the request, and parse the errror if it is past capacity.

Reply:
Great question. 4000 is the current limit as outlined in the [Service Bus service documentation](https://learn.microsoft.com/en-us/azure/service-bus-messaging/batch-delete), yes, which may potentially change over time. It seems this decision of using this const in client is cross language. There are two reasons why SDKs has a hard dependency on this magic number -
1.	Purge API: in purge api implementation, it calls batch-delete multiple times, and the purge API doesn’t need max-messages input from the user. This means client needs to pass _some_ value to service for the required max-messages, this _some_ value must be bounded to 4000 per service contract, a higher value cause purge impl to fail.
2.	Service contract: the batch-delete service call returning 4000 must be treated as there are no more messages left to delete. The Service side feature team is planned to make this the contract when GA-ing. Though they didn’t do it for the beta deployment (which is a bug).